apps/emqx/src/emqx_persistent_session_ds.erl

```diff
@@ -655,6 +657,36 @@ replay_batch(Srs0, Session0, ClientInfo) ->
             Error
     end.
 
+%% Handle `{error, unrecoverable, _}' returned by `enqueue_batch'.
+%% Most likely they mean that the generation containing the messages
+%% has been removed.
+-spec skip_batch(_StreamKey, stream_state(), session(), clientinfo(), _Reason) -> session().
+skip_batch(StreamKey, SRS0, Session = #{s := S0}, ClientInfo, Reason) ->
+    ?SLOG(info, #{
+        msg => "session_ds_replay_unrecoverable_error",
+        reason => Reason,
+        srs => SRS0
+    }),
+    GenEvents = fun
+        F(QoS, SeqNo, LastSeqNo) when SeqNo < LastSeqNo ->
+            FakeMsg = #message{
+                qos = QoS,
+                payload = <<>>
+            },
+            _ = emqx_session_events:handle_event(ClientInfo, {expired, FakeMsg}),
+            F(QoS, inc_seqno(QoS, SeqNo), LastSeqNo);
+        F(_, _, _) ->
+            ok
+    end,
+    %% Treat messages as expired:
+    GenEvents(?QOS_1, SRS0#srs.first_seqno_qos1, SRS0#srs.last_seqno_qos1),
```

Comment:
Aren't `SRS0#srs.first_seqno_qos1` and `SRS0#srs.last_seqno_qos1` equal in case of a error 🤔

Reply:
Not when this function is called during replay (i.e. session got stopped with unacked messages, generation contained these messages expired, session restarts). 
In this case `last_seqno_qosN` contains seqno at the end of the original batch.